blockchain/Validation.go

```diff
@@ -1516,6 +1540,121 @@ var (
 	nTimeTotal     int64
 )
 
+func AcceptToMemoryPoolWorker(config *Config, pool *mempool.Mempool, state *model.ValidationState,
+	tx *model.Tx, limitFree bool, missingInputs *bool, acceptTime int64, txReplaced *list.List,
+	overrideMempoolLimit bool, absurdFee btcutil.Amount, coinsToUncache []*model.OutPoint) bool {
+
+	return true
+}
+
+func AcceptToMemoryPoolWithTime(config *Config, mpool *mempool.Mempool, state *model.ValidationState,
+	tx *model.Tx, limitFree bool, missingInputs *bool, acceptTime int64, txReplaced *list.List,
+	overrideMempoolLimit bool, absurdFee btcutil.Amount) bool {
+
+	coinsToUncache := make([]*model.OutPoint, 0)
+	res := AcceptToMemoryPoolWorker(config, mpool, state, tx, limitFree, missingInputs, acceptTime,
+		txReplaced, overrideMempoolLimit, absurdFee, coinsToUncache)
+
+	if !res {
+		for _, outpoint := range coinsToUncache {
+			coinsTip.UnCache(outpoint)
+		}
+	}
+
+	stateDummy := &model.ValidationState{}
+	FlushStateToDisk(stateDummy, FLUSH_STATE_PERIODIC, 0)
+	return res
+
+}
+func LoadMempool(config *Config) bool {
+	expiryTimeout := (utils.GetArg("-mempoolexpiry", DefaultMempoolExpiry)) * 60 * 60
+
+	fileStr, err := os.OpenFile(conf.GetDataPath()+"/mempool.dat", os.O_RDONLY, 0666)
+	if err != nil {
+		fmt.Println("Failed to open mempool file from disk. Continuing anyway")
+		return false
+	}
+	defer fileStr.Close()
+
+	now := time.Now() // todo C++:nMockTime
+	var count int
+	var failed int
+	var skipped int
+
+	defer func() {
+		if r := recover(); r != nil {
+			fmt.Println("Failed to deserialize mempool data on disk:", err, ". Continuing anyway.")
+		}
+	}()
+
+	// read version firstly
+	version, err := utils.BinarySerializer.Uint32(fileStr, binary.LittleEndian)
+	if err != nil {
+		panic(err)
+	}
+	if version != MempoolDumpVersion {
+		return false
+	}
+
+	num, err := utils.ReadVarInt(fileStr)
+	if err != nil {
+		panic(err)
+	}
+
+	var priorityDummy float64
+	for num > 0 {
+		num--
+		txPoolInfo, err := mempool.DeserializeInfo(fileStr)
+		if err != nil {
+			panic(err)
+		}
+
+		amountDelta := txPoolInfo.FeeDelta
+		if amountDelta != 0 {
+			mpool.PrioritiseTransaction(txPoolInfo.Tx.TxHash(), txPoolInfo.Tx.TxHash().ToString(), priorityDummy, amountDelta)
+		}
+
+		vs := &model.ValidationState{}
+		if txPoolInfo.Time+expiryTimeout > int64(now.Second()) {
+			// todo LOCK(cs_main)
+			AcceptToMemoryPoolWithTime(config, mpool, vs, txPoolInfo.Tx, true, nil,
+				txPoolInfo.Time, nil, false, 0)
+
+			if vs.IsValid() {
+				count++
+			} else {
+				failed++
+			}
+		} else {
+			// timeout
+			skipped++
+		}
+
+		if ShutdownRequested() { // get shutdown signal
+			return false
+		}
+
+		var hash utils.Hash
+		_, err = io.ReadFull(fileStr, hash[:])
+		if err != nil {
+			panic(err)
+		}
+
+		amount, err := utils.BinarySerializer.Uint64(fileStr, binary.LittleEndian)
+		if err != nil {
+			panic(err)
+		}
+		mapDeltas := make(map[utils.Hash]btcutil.Amount)
```

Comment:
the map Serialize error; first : you should write the map element number in the file; second : write each element in the file. 
When you Unserialize the map; first : you should read the map element number; second : you should read each element from the file; three: construct the map struct with each element.